### PR TITLE
Refactor connection v2.0

### PR DESCRIPTION
**Pull request by realDragonium**
Made some changes to make it more procedural the way it handles packets. 

I have tried to make test reusable so that other implementation could also use them.

### CODE REVIEW

File: connection/types.go [hunk 1]
```diff
@@ -28,32 +27,34 @@ const (
 
 // probably needs a better name since its not only used for piping the connection
 type PipeConnection interface {
+	getConn() ByteConnection
```
Comment (haveachin) [inline connection/types.go:30]: Instead of getConn() you could just leave it as conn(). The Go standard library also follows this convention.

File: connection/connection.go [hunk 2]
```diff
@@ -53,170 +53,99 @@ func pipe(c1, c2 PipeConnection) {
 	}
 }
 
-func CreateBasicPlayerConnection(conn Connection, remoteAddr net.Addr) *BasicPlayerConnection {
```
Comment (haveachin) [inline connection/connection.go:56]: I think NewBasicPlayerConn(...) would be less verbose

File: connection/connection.go [hunk 3]
```diff
@@ -53,170 +53,99 @@ func pipe(c1, c2 PipeConnection) {
...
+	c := CreateBasicConnection(conn)
+	return &BasicPlayerConnection{
+		conn:   c,
+		reader: bufio.NewReader(conn),
+		addr:   remoteAddr,
+	}
 }
 
 // Basic implementation of LoginConnection
 type BasicPlayerConnection struct {
```
Comment (haveachin) [inline connection/connection.go:66]: BasicPlayerConn

File: connection/connection.go [hunk 4]
```diff
@@ -53,170 +53,99 @@ func pipe(c1, c2 PipeConnection) {
...
-
-	pk, _ := c.HsPk()
-	var err error
-	c.hs, err = handshaking.UnmarshalServerBoundHandshake(pk)
-	if err != nil {
-		return c.hs, err
-	}
-	c.hasHS = true
-	return c.hs, nil
+func (c *BasicPlayerConnection) HsPk() protocol.Packet {
```
Comment (haveachin) [inline connection/connection.go:89]: HandshakePacket()

File: connection/connection.go [hunk 5]
```diff
@@ -53,170 +53,99 @@ func pipe(c1, c2 PipeConnection) {
...
 
-func (c *BasicPlayerConnection) HsPk() (protocol.Packet, error) {
-	if c.hasHSPk {
-		return c.hsPk, nil
-	}
-	pk, err := c.ReadPacket()
-	if err != nil {
-		return pk, ErrCantGetHSPacket
-	}
+func (c *BasicPlayerConnection) SetHsPk(pk protocol.Packet) {
```
Comment (haveachin) [inline connection/connection.go:93]: SetHandshakePacket(...)

Comment (haveachin) [reply connection/connection.go:93]: I would also change the receiver parameter name to conn instead of just c

File: connection/types.go [hunk 6]
```diff
@@ -28,32 +27,34 @@ const (
 
 // probably needs a better name since its not only used for piping the connection
 type PipeConnection interface {
```
Comment (haveachin) [inline connection/types.go:29]: PipeConn

File: connection/types.go [hunk 7]
```diff
@@ -28,32 +27,34 @@ const (
...
+	getConn() ByteConnection
+}
+
+type ByteConnection interface {
```
Comment (haveachin) [inline connection/types.go:33]: ByteConn

File: connection/types.go [hunk 8]
```diff
@@ -28,32 +27,34 @@ const (
...
 	io.Writer
 	io.Reader
+	io.Closer
 }
 
 type Connection interface {
 	infrared.PacketWriter
 	infrared.PacketReader
-	PipeConnection
 }
```
Comment (haveachin) [inline connection/types.go:42]: Maybe we should just move the PacketWriter and PacketReader over from the old Conn and delete it

File: connection/types.go [hunk 9]
```diff
@@ -28,32 +27,34 @@ const (
...
 
 type HSConnection interface {
```
Comment (haveachin) [inline connection/types.go:44]: HandshakeConn

File: connection/types.go [hunk 10]
```diff
@@ -28,32 +27,34 @@ const (
...
 	Connection
-	RemoteAddressConnection
-	Handshake() (handshaking.ServerBoundHandshake, error)
-	HsPk() (protocol.Packet, error)
-}
+	Handshake() handshaking.ServerBoundHandshake
+	HsPk() protocol.Packet
```
Comment (haveachin) [inline connection/types.go:47]: HandshakePacket

File: connection/types.go [hunk 11]
```diff
@@ -28,32 +27,34 @@ const (
...
+
+	SetHsPk(pk protocol.Packet)
```
Comment (haveachin) [inline connection/types.go:49]: SetHandshakePacket

File: connection/types.go [hunk 12]
```diff
@@ -28,32 +27,34 @@ const (
...
+	SetHandshake(hs handshaking.ServerBoundHandshake)
 
-type GatewayConnection interface {
-	RemoteAddressConnection
-	ServerAddr() string
+	RemoteAddr() net.Addr
 }
 
 type LoginConnection interface {
```
Comment (haveachin) [inline connection/types.go:55]: LoginConn

File: connection/types.go [hunk 13]
```diff
@@ -28,32 +27,34 @@ const (
...
 	HSConnection
-	Name() (string, error)
-	LoginStart() (protocol.Packet, error) // Need more work
+	PipeConnection
 }
 
 type StatusConnection interface {
```
Comment (haveachin) [inline connection/types.go:60]: StatusConn

File: connection/types.go [hunk 14]
```diff
@@ -62,10 +63,5 @@ type StatusConnection interface {
 
 type ServerConnection interface {
```
Comment (haveachin) [inline connection/types.go:64]: ServerConn

File: gateway/gateway.go [hunk 15]
```diff
@@ -2,33 +2,34 @@ package gateway
...
 )
 
 var (
+	ErrCantGetHSPacket   = errors.New("cant get handshake packet from caller")
 	ErrNoServerFound     = errors.New("no server was found for this request")
 	ErrNotValidHandshake = errors.New("this connection didnt provide a valid handshake")
 )
 
-func CreateBasicGatewayWithStore(store ServerStore, ch <-chan connection.GatewayConnection) BasicGateway {
+func CreateBasicGatewayWithStore(store ServerStore, ch <-chan connection.HSConnection) BasicGateway {
```
Comment (haveachin) [inline gateway/gateway.go:17]: NewBasicGatewayWithStore(...)

File: gateway/gateway.go [hunk 16]
```diff
@@ -2,33 +2,34 @@ package gateway
...
 }
 
 type ServerData struct {
-	ConnCh chan<- connection.GatewayConnection
+	ConnCh chan<- connection.HSConnection
 }
 
 type Gateway interface {
-	HandleConnection(conn connection.GatewayConnection) error
+	HandleConnection(conn connection.HSConnection) error
```
Comment (haveachin) [inline gateway/gateway.go:27]: HandleConn(...)

File: gateway/gateway.go [hunk 17]
```diff
@@ -38,14 +39,25 @@ func (g *BasicGateway) Start() error {
 	}
 }
 
-func (g *BasicGateway) handleConnection(conn connection.GatewayConnection) error {
-	addr := conn.ServerAddr()
-	log.Printf("[i] %s requests proxy for address %s", conn.RemoteAddr(), addr)
+func (g *BasicGateway) handleConnection(conn connection.HSConnection) error {
```
Comment (haveachin) [inline gateway/gateway.go:42]: (gateway *BasicGateway) handleConn(...)

File: gateway/listener.go [hunk 18]
```diff
@@ -14,7 +14,7 @@ var (
 
 type OuterListener interface {
 	Start() error
-	Accept() (connection.Connection, net.Addr)
+	Accept() (net.Conn, net.Addr)
 }
 
 func CreateBasicOuterListener(addr string) OuterListener {
```
Comment (haveachin) [inline gateway/listener.go:20]: NewBasicOuterListener(...)

File: protocol/packet_test.go [hunk 19]
```diff
@@ -1,25 +1,28 @@
-package protocol
+package protocol_test
```
Comment (haveachin) [inline protocol/packet_test.go:1]: Why did you change the package name

**Review by haveachin — COMMENTED** [review]
LGTM 🚀
I just added a few comments and suggestions for renaming.

File: connection/types.go (hunk 1 above)
Comment (realDragonium) [reply connection/types.go:30]: im aware of the convention but i keep forgetting about it  ':)

File: connection/connection.go (hunk 2 above)
Comment (realDragonium) [reply connection/connection.go:56]: Yea the New is better than the Create, good one

File: connection/connection.go (hunk 5 above)
Comment (realDragonium) [reply connection/connection.go:93]: with all those conns its probably getting harder to know which one is what. if we were to do that, we should probably rename the conn for net.Conn to netConn and the conn for ByteConn to byteConn

File: protocol/packet_test.go (hunk 19 above)
Comment (realDragonium) [reply protocol/packet_test.go:1]: I think its a good convention to start doing it since that way we will only test methods/functions which are exposed which should help with creating feature tests instead of implementation tests. Besides that packages with _test behind will be build in their own binary. so I didnt really saw a downside to do it.

Comment (haveachin) [reply protocol/packet_test.go:1]: Nice, I wasn't aware of the binary part

File: connection/connection.go (hunk 5 above)
Comment (haveachin) [reply connection/connection.go:93]: Sounds like a good idea

**Review by haveachin — APPROVED** [review]
